src/libraries/Microsoft.PowerFx.Interpreter/Values/QueryableTableValue.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Core.IR;
+
+namespace Microsoft.PowerFx.Types
+{
+    public abstract class QueryableTableValue : TableValue
+    {
+        private Task<List<DValue<RecordValue>>> _taskRows;
+        private readonly object _lock = new object();
+
+        internal QueryableTableValue(IRContext irContext)
+            : base(irContext)
+        {
+        }
+
+        internal abstract TableValue Filter(LambdaFormulaValue lambda);
+
+        public bool HasCachedRows
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _taskRows != null;
+                }
+            }
+        }
+
+        protected abstract Task<List<DValue<RecordValue>>> GetRowsAsync();
+
+        public sealed override IEnumerable<DValue<RecordValue>> Rows
+        {
+            get
+            {
+                // Not the best, but doesn't require major breaking changes to TableValue
+                lock (_lock)
+                {
+                    if (_taskRows == null)
+                    {
+                        _taskRows = GetRowsAsync();
```

Comment:
>GetRowsAsync [](http://example.com/codeflow?start=36&length=12)

You can't call this under a lock. 
Call in a local and then just assign under the lock - possible that it may run twice, but that's uinevitable.

Reply:
Yeah I wasn't super sure about this one...do you recommend using `Lazy<ReadOnlyList<RecordValue>>` instead?

Reply:
Switched to `Lazy<Task...>` so it is only initialized once and needs no lock